pkg/controller/ttl/ttlcontroller.go

```diff
@@ -0,0 +1,237 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ttl
+
+import (
+	"fmt"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	listers "k8s.io/kubernetes/pkg/client/listers/core/v1"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/informers"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// ObjectTTLAnnotation is a suggestion for kubelet for how long it can cache
+	// an object (e.g. secret, config map) before fetching from apiserver again.
+	ObjectTTLAnnotation = "node.kubernetes.io/ttl"
+)
+
+type TTLController struct {
+	kubeClient clientset.Interface
+
+	// nodeStore is a local cache of nodes.
+	nodeStore listers.NodeLister
+
+	// Secrets that need to be synced.
+	queue workqueue.RateLimitingInterface
+
+	// Returns true if all underlying informers are synced.
+	hasSynced func() bool
+
+	// Number of nodes in the cluster.
+	// Should be used via atomic operations on it.
+	nodeCount int32
+}
+
+func NewTTLController(nodeInformer informers.NodeInformer, kubeClient clientset.Interface) *TTLController {
+	ttlc := &TTLController{
+		kubeClient: kubeClient,
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ttlcontroller"),
+	}
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ttlc.addNode,
+		UpdateFunc: ttlc.updateNode,
+		DeleteFunc: ttlc.deleteNode,
+	})
+
+	ttlc.nodeStore = listers.NewNodeLister(nodeInformer.Informer().GetIndexer())
+	ttlc.hasSynced = nodeInformer.Informer().HasSynced
+
+	return ttlc
+}
+
+func (ttlc *TTLController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer ttlc.queue.ShutDown()
+
+	glog.Infof("Starting TTL controller")
+	defer glog.Infof("Shutting down TTL controller")
+	if !cache.WaitForCacheSync(stopCh, ttlc.hasSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(ttlc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (ttlc *TTLController) addNode(obj interface{}) {
+	node, ok := obj.(*v1.Node)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("unexpected object types: %v", obj))
+		return
+	}
+	atomic.AddInt32(&ttlc.nodeCount, 1)
```

Comment:
That's not true. There might be some periods where some nodes have different ttls than other nodes, but those should be very short periods.
Note that Kubelet is updating status every 10 seconds. As a result, we will be processing each node every 10s. That means, that once the cluster changes the size to from one group to the other, within 10s or so, all nodes should have updated ttls.
I think this is good enough.

I'm adding a comment to clarify this.